internal/pkg/scalable/deployments.go

```diff
@@ -3,48 +3,89 @@ package scalable
 import (
 	"context"
 	"fmt"
+	"log/slog"
+	"math"
 
 	appsv1 "k8s.io/api/apps/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
-	"k8s.io/client-go/kubernetes"
 )
 
 // getDeployments is the getResourceFunc for Deployments
-func getDeployments(namespace string, clientset *kubernetes.Clientset, ctx context.Context) ([]Workload, error) {
+func getDeployments(namespace string, clientsets *Clientsets, ctx context.Context) ([]Workload, error) {
 	var results []Workload
-	deployments, err := clientset.AppsV1().Deployments(namespace).List(ctx, metav1.ListOptions{TimeoutSeconds: &timeout})
+	deployments, err := clientsets.Kubernetes.AppsV1().Deployments(namespace).List(ctx, metav1.ListOptions{TimeoutSeconds: &timeout})
 	if err != nil {
 		return nil, fmt.Errorf("failed to get deployments: %w", err)
 	}
 	for _, item := range deployments.Items {
-		results = append(results, deployment{&item})
+		results = append(results, &deployment{&item})
 	}
 	return results, nil
 }
 
-// deployment is a wrapper for appsv1.Deployment to implement the scalableResource interface
+// deployment is a wrapper for appsv1.Deployment to implement the Workload interface
 type deployment struct {
 	*appsv1.Deployment
 }
 
-// SetReplicas sets the amount of replicas on the resource. Changes won't be made on kubernetes until update() is called
-func (d deployment) SetReplicas(replicas int) {
+// setReplicas sets the amount of replicas on the resource. Changes won't be made on kubernetes until update() is called
+func (d *deployment) setReplicas(replicas int) error {
+	if replicas > math.MaxInt32 || replicas < 0 {
+		return errBoundOnScalingTargetValue
+	}
+
+	// #nosec G115
 	newReplicas := int32(replicas)
 	d.Spec.Replicas = &newReplicas
+	return nil
 }
 
-// GetCurrentReplicas gets the current amount of replicas of the resource
-func (d deployment) GetCurrentReplicas() (int, error) {
+// getCurrentReplicas gets the current amount of replicas of the resource
+func (d *deployment) getCurrentReplicas() (int, error) {
 	replicas := d.Spec.Replicas
 	if replicas == nil {
 		return 0, errNoReplicasSpecified
 	}
 	return int(*d.Spec.Replicas), nil
 }
 
+// ScaleUp upscale the resource
+func (d *deployment) ScaleUp() error {
+	originalReplicas, err := getOriginalReplicas(d)
+	if err != nil {
+		return fmt.Errorf("failed to get original replicas for workload: %w", err)
+	}
+	if originalReplicas == nil {
+		slog.Debug("original replicas is not set, skipping", "workload", d.GetName(), "namespace", d.GetNamespace())
+		return nil
+	}
+
+	err = d.setReplicas(*originalReplicas)
+	if err != nil {
+		return fmt.Errorf("failed to set original replicas for workload: %w", err)
+	}
+	removeOriginalReplicas(d)
+	return nil
+}
+
+// ScaleDown downscale the resource
+func (d *deployment) ScaleDown(downscaleReplicas int) error {
```

Comment:
I just realised we need to change something back here, currently the ScaleDown with original replicas would set the orginal replicas to the downscaled replicas in the second scan being downscaled. For all resources that use orignial replicas, we need to add back the check if CurrentReplicas == DownscaleReplicas